apollo-federation/src/connectors/json_selection/methods/public/eq.rs

```diff
@@ -28,9 +29,32 @@ fn eq_method(
     if let Some(MethodArgs { args, .. }) = method_args {
         if let [arg] = args.as_slice() {
             let (value_opt, arg_errors) = arg.apply_to_path(data, vars, input_path);
-            let matches = value_opt.is_some_and(|value| &value == data);
+            let mut apply_to_errors = arg_errors;
+            let matches = value_opt.and_then(|value| match (data, &value) {
+                // Number comparisons: Always convert to float so 1 == 1.0
+                (JSON::Number(left), JSON::Number(right)) => {
+                    let left = match number_value_as_float(left, method_name, input_path) {
+                        Ok(f) => f,
+                        Err(err) => {
+                            apply_to_errors.push(err);
+                            return None;
+                        }
+                    };
+                    let right = match number_value_as_float(right, method_name, input_path) {
+                        Ok(f) => f,
+                        Err(err) => {
+                            apply_to_errors.push(err);
+                            return None;
+                        }
+                    };
 
-            return (Some(JSON::Bool(matches)), arg_errors);
+                    Some(JSON::Bool(left == right))
+                }
+                // Everything else
+                _ => Some(JSON::Bool(&value == data)),
```

Comment:
Since we're not erroring/warning at runtime, in some sense there's no error/warning to anticipate at shape processing time. I still think it could be helpful to warn in the `{ne,eq}_shape` functions, but we can go either way, depending on how confusing this ends up being for developers.